lib/discourse_auth_cookie.rb

```diff
@@ -0,0 +1,137 @@
+
+class DiscourseAuthCookie
+  class InvalidCookie < StandardError; end
+
+  class Encryptor
+    def encrypt_and_sign(plain_cookie)
+      encryptor.encrypt_and_sign(plain_cookie)
+    end
+
+    def decrypt_and_verify(cipher_cookie)
+      encryptor.decrypt_and_verify(cipher_cookie)
+    rescue ActiveSupport::MessageVerifier::InvalidSignature
+      raise InvalidCookie.new
+    end
+
+    private
+
+    def encryptor
+      @encryptor ||= begin
+        key = Rails.application.key_generator.generate_key("discourse-auth-cookie", 32)
+        ActiveSupport::MessageEncryptor.new(
+          key,
+          key,
+          cipher: "aes-256-cbc",
+          digest: "SHA256"
+        )
+      end
+    end
+  end
+
+  TOKEN_SIZE ||= 32
+
+  TOKEN_KEY ||= "token"
+  ID_KEY ||= "id"
+  TL_KEY ||= "tl"
+  TIME_KEY ||= "time"
+  VALID_KEY ||= "valid"
+  private_constant *%i[
+    TOKEN_KEY
+    ID_KEY
+    TL_KEY
+    TIME_KEY
+    VALID_KEY
+  ]
+
+  attr_reader *%i[token user_id trust_level timestamp valid_for]
+
+  def self.parse(raw_cookie, secret = Rails.application.secret_key_base)
+    # v0 of the cookie was simply the auth token itself. we need this for
+    # backward compatibility so we don't wipe out existing sessions.
+    # TODO: drop this line (and a test case in
+    # default_current_user_provider_spec.rb) in Nov 2023
+    return new(token: raw_cookie) if raw_cookie.size == TOKEN_SIZE
+
+    data = Encryptor.new.decrypt_and_verify(raw_cookie)
+
+    token = nil
+    user_id = nil
+    trust_level = nil
+    timestamp = nil
+    valid_for = nil
+
+    data.split(",").each do |part|
+      prefix, val = part.split(":", 2)
+      val = val.presence
+      if prefix == TOKEN_KEY
+        token = val
+      elsif prefix == ID_KEY
+        user_id = val
+      elsif prefix == TL_KEY
+        trust_level = val
+      elsif prefix == TIME_KEY
+        timestamp = val
+      elsif prefix == VALID_KEY
+        valid_for = val
+      end
+    end
+
+    new(
+      token: token,
+      user_id: user_id,
+      trust_level: trust_level,
+      timestamp: timestamp,
+      valid_for: valid_for,
+    )
+  end
+
+  def self.validate_signature!(data, sig, secret)
+    data = data.to_s
+    sig = sig.to_s
+    if compute_signature(data, secret) != sig
+      raise InvalidCookie.new
+    end
+  end
+
+  def self.compute_signature(data, secret)
+    OpenSSL::HMAC.hexdigest("sha256", secret, data)
+  end
+
+  def initialize(token:, user_id: nil, trust_level: nil, timestamp: nil, valid_for: nil)
+    @token = token
+    @user_id = user_id.to_i if user_id
+    @trust_level = trust_level.to_i if trust_level
+    @timestamp = timestamp.to_i if timestamp
+    @valid_for = valid_for.to_i if valid_for
+  end
+
+  def to_text
+    parts = []
+    parts << [TOKEN_KEY, token].join(":")
+    parts << [ID_KEY, user_id].join(":")
+    parts << [TL_KEY, trust_level].join(":")
+    parts << [TIME_KEY, timestamp].join(":")
+    parts << [VALID_KEY, valid_for].join(":")
```

Comment:
Creating an array and running `join` feels a little wasteful here. For example, `[TOKEN_KEY, token].join(":")` can just be "#{TOKEN_KEY}:#{token}".